server/apps/publish/archive_publish.py

```diff
@@ -301,33 +304,31 @@ def process_takes(self, updates_of_take_to_be_published, package_id, original_of
                     sequence_num_of_take_to_be_published = r[SEQUENCE]
                     break
 
-            if sequence_num_of_take_to_be_published:
-                if self.published_state != "killed":
-                    for sequence in range(sequence_num_of_take_to_be_published, 0, -1):
-                        previous_take_ref = next(ref for ref in take_refs if ref.get(SEQUENCE) == sequence)
-                        if previous_take_ref[GUID_FIELD] != take_article_id:
-                            previous_take = super().find_one(req=None, _id=previous_take_ref[GUID_FIELD])
-
-                            if not previous_take:
-                                raise PublishQueueError.article_not_found_error(
-                                    Exception("Take with id %s not found" % previous_take_ref[GUID_FIELD]))
+            if takes and self.published_state != 'killed':
+                body_html_list = [take.get('body_html', '') for take in takes]
+                if self.published_state == 'published':
+                    body_html_list.append(body_html)
+                else:
+                    body_html_list[sequence_num_of_take_to_be_published - 1] = body_html
 
-                            if previous_take and previous_take[config.CONTENT_STATE] not in ['published', 'corrected']:
-                                raise PublishQueueError.previous_take_not_published_error(
-                                    Exception("Take with id {} is not published in Takes Package with id {}"
-                                              .format(previous_take_ref[GUID_FIELD], package[config.ID_FIELD])))
+                package_updates['body_html'] = '<br>'.join(body_html_list)
+            else:
+                package_updates['body_html'] = body_html
 
-                            package_updates['body_html'] = \
-                                previous_take['body_html'] + '<br>' + package_updates['body_html']
+            metadata_tobe_copied = ['headline', 'abstract', 'anpa_category', 'pubstatus', 'slugline', 'urgency',
```

Comment:
Ah, I now see this is used below to check which fields should be copied (`for metadata in metadata_tobe_copied...`). In that case `set` is a better data structure to use for this purpose, because the `in` operator takes a constant time, which is faster than the `list`'s linear time.